### ISSUE TITLE
Support/ignore `export as namespace` TypeScript syntax

### ISSUE THREAD

**Issue by shroudedcode** [evanw/esbuild#2070]
Recently I've come across multiple packages which use the (admittedly poorly documented)[ `export as namespace` syntax](https://stackoverflow.com/a/44849515) in their declaration files to declare that they provide both CommonJS exports _and_ a global namespace object. [Searching DefinitelyTyped alone](https://grep.app/search?q=export%20as%20namespace%20&filter[lang][0]=TypeScript&filter[repo][0]=DefinitelyTyped/DefinitelyTyped) brings up 706 occurrences of this pattern.

When bundling a script importing one of these packages ESBuild reports the following error:

```js
✘ [ERROR] Unexpected "as"

    node_modules/@types/mapbox__polyline/index.d.ts:15:7:
      15 │ export as namespace polyline;
         ╵        ~~
```

It seems like ESBuild simply doesn't know the `export as namespace` syntax yet, so the `as` keyword is flagged as unexpected in that context.

## Proposal

Since (as far as I know) ESBuild only supports actual module formats (and not global namespace objects), I think it would be a good enough solution to just completely ignore that line. Modules containing it would then be treated solely according to their remaining export declarations.

**Comment by magic-akari** [comment]
It is weird.
IIRC, esbuild doesn't read `.d.ts`.

**Comment by shroudedcode** [comment]
@magic-akari 

> IIRC, esbuild doesn't read `.d.ts`.

I thought so too, but [this comment from the changelog](https://github.com/evanw/esbuild/blob/2b644d6bbf120833281e041caabfbd12dcf533d4/CHANGELOG.md#:~:text=This%20fixes%20this%20case%20because%20.d.ts%20files%20typically%20end%20up%20being%20empty%20after%20esbuild%20parses%20them%20since%20they%20typically%20only%20contain%20type%20declarations.) seems to suggest the opposite.

**Comment by shroudedcode** [comment]
Thanks for the help @magic-akari and @evanw!

I've spent some more time looking into this issue and realized that it was partly caused by the following setting in my project's `tsconfig.json`:

```
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@mapbox/polyline": ["node_modules/@types/mapbox__polyline/index.d.ts"]
    }
  }
}
```

I initially set this up to work around an issue with TypeScript's experimental `NodeNext` module resolution mode which seems to have been fixed in the meantime. What I didn't know at that time was that ESBuild respects the `paths` option as well but for resolving actual source files.

Maybe the real fix here isn't to add support for `export as namespace`, but instead to ignore any `paths` entries ending in `.d.ts` because those are only added to fix typing issues. 🤔

I've created a little reproduction repo (with regular TypeScript module resolution because that doesn't seem to be part of the issue) [here](https://github.com/shroudedcode/esbuild-namespace-issue-repro) which has two scripts (`bundle` and `bundle:paths`) which bundle the `index.ts` file once with and once without the `paths` option.
